apps/prairielearn/src/lib/assets.js

```diff
@@ -1,70 +1,90 @@
 // @ts-check
 const crypto = require('crypto');
+const express = require('express');
 const path = require('path');
-const fs = require('fs');
 const { hashElement } = require('folder-hash');
-const { APP_ROOT_PATH } = require('./paths');
+const compiledAssets = require('@prairielearn/compiled-assets');
 
-let assetsHash = 'NOT_SET';
-let elementsHash = 'NOT_SET';
-const cachedPackageVersionHashes = {};
+const { config } = require('./config');
+const { REPOSITORY_ROOT_PATH, APP_ROOT_PATH } = require('./paths');
+const staticNodeModules = require('../middlewares/staticNodeModules');
+const elementFiles = require('../pages/elementFiles/elementFiles');
+
+let assetsPrefix = null;
 
 /**
- * Computes the hash of the given directory and returns the first 16 characters.
+ * Computes the hash of the given path and returns the first 16 characters.
+ * The path can be a file or a directory.
  *
- * @param {string} dir
+ * @param {string} pathToHash
  * @returns {Promise<string>}
  */
-async function hashDirectory(dir) {
-  const { hash } = await hashElement(dir, { encoding: 'hex' });
+async function hashPath(pathToHash) {
+  const { hash } = await hashElement(pathToHash, { encoding: 'hex' });
   return hash.slice(0, 16);
 }
 
-async function computeAssetsHash() {
-  assetsHash = await hashDirectory(path.join(APP_ROOT_PATH, 'public'));
-}
-
-async function computeElementsHash() {
-  elementsHash = await hashDirectory(path.join(APP_ROOT_PATH, 'elements'));
-}
-
-function getPackageVersion(packageName) {
-  try {
-    return require(`${packageName}/package.json`).version;
-  } catch (e) {
-    if (e.code !== 'ERR_PACKAGE_PATH_NOT_EXPORTED') {
-      throw e;
-    }
-
-    // If we can't directly resolve the package's `package.json` file, we'll
-    // `require.resolve` the package itself, and then do a little manipulation
-    // to get the `package.json` path from there.
-
-    // Get the resolved path to the package entrypoint, which will look something
-    // like `/absolute/path/to/node_modules/package-name/index.js`.
-    const pkgPath = require.resolve(packageName);
-
-    // Strip off everything after the last `/node_modules/`, then append the
-    // package name.
-    const nodeModulesToken = '/node_modules/';
-    const lastNodeModulesIndex = pkgPath.lastIndexOf(nodeModulesToken);
-    const pkgJsonPath = path.resolve(
-      pkgPath.slice(0, lastNodeModulesIndex + nodeModulesToken.length),
-      packageName,
-      'package.json'
-    );
-
-    const pkgJson = JSON.parse(fs.readFileSync(pkgJsonPath, 'utf-8'));
-    return pkgJson.version;
-  }
+async function computeCachebuster() {
+  const hashes = await Promise.all([
+    hashPath(path.join(APP_ROOT_PATH, 'public')),
+    hashPath(path.join(APP_ROOT_PATH, 'elements')),
+    hashPath(path.join(REPOSITORY_ROOT_PATH, 'yarn.lock')),
```

Comment:
The assumption here is that the contents of our `node_modules` directories change iff `yarn.lock` changes. I believe that should be a valid assumption in practice.

Reply:
Do we also need `build` here? Or are we storing this in `public/build` now? It feels a bit weird to me to put `build` inside `public`?

Reply:
We can certainly put `build` somewhere else, but conceptually it's public, so it seemed to me to be a reasonable place to put it. (Note that PT uses `public/build` too.)

FWIW, even if `build` was not located in `public`, we wouldn't need it to factor into the cachebuster, as each file in `build` is guaranteed to have a fingerprint embedded in its name.